url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/C1541Chips.java

```diff
@@ -0,0 +1,962 @@
+    init(cpu);
+  }
+
+  public void initIEC2(C64Screen s) {
+    cia2 = s;
+  }
+
+  public void setReader(C64Reader reader) {
+    log("Setting reader...");
+    this.reader = reader;
+    reader.setDiskListener(this);
+  }
+
+  public final int performRead(int address, long cycles) {
+    switch (address) {
+    case 0x1800: // VIA1 PB - IEEE Serial Bus / IEC
+      return (via1PB & 0x1a
+          | ((iecLines & cia2.iecLines) >> 7) & 0x01    // DATA
+          | ((iecLines & cia2.iecLines) >> 4) & 0x04    // CLK
+          | (cia2.iecLines << 3) & 0x80) ^ 0x85;        // ATN
+    case 0x1801: // VIA1 PA
+    case 0x180f: // VIA1 PA
+      via1IFlag &= ~2;
+      checkInterrupt(1, "read from pa");
+      return 0xff; // via1PA;
+    case 0x1802: // VIA1 CB
+      return via1CB;
+    case 0x1803: // VIA1 CA
+      return via1CA;
+    case 0x1804:
+      via1IFlag &= 0xbf;
+      checkInterrupt(1, "read T1 low");
+      return via1T1Ctr & 0xff;
+    case 0x1805:
+      return via1T1Ctr >> 8;
+          case 0x1806:
+            return via1T1Latch & 0xff;
+          case 0x1807:
+            return via1T1Latch >> 8;
+          case 0x1808:
+            via1IFlag &= 0xdf;
+            checkInterrupt(1, "read T2 low");
+            return via1T2Ctr & 0xff;
+          case 0x1809:
+            return via1T2Ctr >> 8;
+          case 0x180a:
+            return via1SerialRegister;
+          case 0x180b:
+            return via1AuxControl;
+          case 0x180c:
+            return via1PerControl;
+          case 0x180d:
+            return via1IFlag;
+          case 0x180e:
+            return via1IEnable;
+
+
+          case 0x1c00: // VIA2 PB
+            return (via2PB & 0x6f) | sync() | writeProtect();
+          case 0x1c01: // VIA2 PA - read from disk!
+          case 0x1c0f:
+            return readByte();
+          case 0x1c02: // VIA2 CB
+            return via2CB;
+          case 0x1c03: // VIA2 CA
+            return via2CA;
+          case 0x1c04:
+            via2IFlag &= 0xbf;
+            checkInterrupt(1, "read T1 low");
+            return via2T1Ctr & 0xff;
+          case 0x1c05:
+            return via2T1Ctr >> 8;
+          case 0x1c06:
+            return via2T1Latch & 0xff;
+          case 0x1c07:
+            return via2T1Latch >> 8;
+          case 0x1c08:
+            via2IFlag &= 0xdf;
+//          System.out.println("Updated via2IFlag: " +
+//          Integer.toHexString(via2IFlag));
+            checkInterrupt(2, "read T2 low");
+            return via2T2Ctr & 0xff;
+          case 0x1c09:
+            return via2T2Ctr >> 8;
+          case 0x1c0a:
+            return via2SerialRegister;
+          case 0x1c0b:
+            return via2AuxControl;
+          case 0x1c0c:
+            return via2PerControl;
+          case 0x1c0d:
+            return via2IFlag;
+          case 0x1c0e:
+            return via2IEnable;
+
+    }
+
+    return 0;
+  }
+
+  public final void performWrite(int address, int data, long cycles) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 134 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.